### PR TITLE
Allow purging to be toggled

### PR DESCRIPTION
**Pull request by gingermusketeer**

### CODE REVIEW

File: test/fastly-rails/configuration_test.rb [hunk 1]
```diff
@@ -65,6 +65,26 @@
...
+    end
+
+    it 'is true' do
+      assert_equal true, configuration.cache_headers_enabled?
+    end
+  end
+
+  describe 'cache_headers_enabled?' do
+    before do
+      configuration.purging_enabled = true
```
Comment (thommahoney) [inline test/fastly-rails/configuration_test.rb]: This is the default behavior. Could you also add tests for the negative? (same for `cache_headers_enabled` please)

File: test/fastly-rails_test.rb [hunk 2]
```diff
@@ -38,6 +38,8 @@
         c.password  = password
         c.max_age   = max_age
         c.service_id = service_id
+        c.cache_headers_enabled = true
```
Comment (thommahoney) [inline test/fastly-rails_test.rb]: These lines should be unnecessary as `true` is the default.

File: test/fastly-rails_test.rb [hunk 3]
```diff
@@ -47,10 +49,27 @@
...
+      assert_equal true, configuration.cache_headers_enabled?
+      assert_equal true, configuration.purging_enabled?
     end
 
     it 'should return a valid client' do
       assert_instance_of FastlyRails::Client, client
     end
   end
+
+  describe 'purge_by_key' do
```
Comment (thommahoney) [inline test/fastly-rails_test.rb]: It would be best to also test that the method does not execute `client.purge_by_key` when `purging_enabled` is `false`.

**Comment by thommahoney** [comment]
@gingermusketeer Thanks for your PR! I've made a few comments asking for some tests. Looks good otherwise.

**Comment by gingermusketeer** [comment]
@thommahoney I have made the requested changes. Let me know if there is anything else you want changed

**Comment by gingermusketeer** [comment]
@thommahoney Rebased

**Comment by aspires** [comment]
cc @fastly/app

**Comment by set5think** [comment]
@gingermusketeer Hi, I had a couple questions/points of feedback about this PR.

My main question is: what is your use-case for this?  From the examples you added in the README, and what I can infer, this seems like a solution to difficulties you were having during local development; but I want to be sure I understand the problem.

The reason I ask is because the plugin doesn't work unless the application using it explicitly calls the methods provided by the plugin.  So, adding code that toggles whether the methods work in the plugin feels like the wrong place to be doing that.  The disadvantage to adding code at the plugin-level (or maybe the current implementation) is that you still call the methods, it just returns before anything happens.  At the application layer, you would simply not call those methods, whether it's through an `unless` or `if` guard on a `before_filter` or an activerecord callback, or however else someone would have it implemented.

More importantly, even though it seems like you could do a separation between the purging methods and the cache-headers, having two separate toggles for those can definitely lead to problems.  If you have purging enabled, but header-setting disabled, nothing will ever get purged, because you never set surrogate-keys on any of your requests.

If you could talk through the main problem you'd like to see solved, we definitely would love to help get that into this codebase.  Seeing that this has been open for a while, I'll definitely get back to you very quickly about this; as it's now on our radar.

Thanks either way; I'm hoping this will lead to the right resolution for current and future users of this plugin.

**Comment by adamcohen** [comment]
This change would allow us to use the purging/header methods as is, but override them at the plugin level, rather than necessitating adding a conditional when using them.  I'm using fastly-rails in a large number of projects, and I often want to toggle this behaviour in dev mode, so I think the plugin is the ideal place to add this.

**Comment by gingermusketeer** [comment]
@set5think the main use case for this is to allow purging the cache to be 
disabled when running the app locally. The alternative to this is 
- Setting the config locally so fastly-rails does not raise an exception when purging
- Wrapping each purge call in a `if Rails.env.development?` as you have 
suggested

My thinking when submitting this PR was that there was an existing PR #32 which 
did what i thought was needed and just needed a bit of refinement. So i took on
doing the work needed so as to help get it merged. I think having this as part of the gem will make it a lot nicer to use. I would have thought anyone using this would have this problem

**Comment by set5think** [comment]
@gingermusketeer @adamcohen Thanks for the information.  A couple thoughts:

- What problems do you run into with leaving the header functionality intact?  Especially considering you can pass the values you want via configuration (and you could definitely change the value based on environment), and the default values seem local-development-friendly.  Perhaps phrased differently, would solely disabling purging so that the request to Fastly doesn't occur be sufficient?

The reason I ask is because it seems like a better place to make a `development-or-test mode` change would be at the `FastlyRails::Client` layer.  We should just have a mockable client so that we don't actually try to make external requests given a configuration request, but you still are able to test behavior that you expect.

At the least, we definitely need to collapse these two configurations into one, because as I mentioned in my first response, having them separately makes it easy to break purging functionality in the application.  But, I do think changing `FastlyRails::Client` is a more optimal solution.  We'll probably still need configuration to either disable/enable external requests, or force external requests in development or test  mode (meaning we'd say that in those two environments, we assume a mocked environment by default).

**Comment by set5think** [comment]
> Wrapping each purge call in a if Rails.env.development? as you have suggested

Have you considered using an ActiveSupport::Concern to consolidate FastlyRails-specific functionality in one place, to possibly make it easier to change behavior like this?

**Comment by gingermusketeer** [comment]
@set5think I have not had any issues with it myself. It was part of the original PR so i left it in. I think when developing locally it would be best to have no caching as it will only lead to frustration.

I made them two separate flags for this reason. You probably will not need to disable the headers but if it causes an issue you can.

I like the idea of pushing the purging down to the client. Are you going to tackle this?

**Comment by set5think** [comment]
@gingermusketeer - the default values for the cache headers should have no impact on local development.  It's actually not named perfectly well; as the default values are really geared towards letting Fastly cache the data, not your browser.  Additionally, you can change the values of the headers using configuration for your dev/test env conditionally, if you need to.

I will look into the client changes; give me a day to figure out when I can get to it.  If I can't get to it soon, I would say we can look at a way of getting this PR merged, but I do think it will require collapsing the two configuration parameters to one, so it either disables both items simultaneously, or enables them simultaneously.  Alternatively, we don't include header toggling, because there's an alternative solution to it and the default values should not cause local development issues.

**Comment by set5think** [comment]
@gingermusketeer really sorry about the delay on this;  I've gotten much busier over the last week.  There's no way I can get to making changes to this any time soon.

I'd like to get this PR merged; can you select from either of the options I presented (either collapse the configuration into one, or remove header toggling)?  I'm impartial to which is selected.

If you can take care of that, and the build passes; I will go ahead and merge this immediately.

Thanks!

**Comment by gingermusketeer** [comment]
@set5think I have updated the PR but for some reason the build is failing. I tried to run a rebuild but could not work out how to do it :(

**Comment by adamcohen** [comment]
@gingermusketeer @set5think any progress on this? I'd love to see this merged, would be really useful for me in development mode.  Thanks

**Comment by set5think** [comment]
I'll look into this today

**Comment by gingermusketeer** [comment]
Any update on this PR?

**Comment by set5think** [comment]
I'm not entirely sure why these builds started failing, but it certainly doesn't have to do with this PR.  I've run the tests on this PR and everything is passing; the resolution to the build failure will be done in a separate PR.
